Sources/SemanticAnalyzer/SemanticAnalyzer.swift

```diff
@@ -402,7 +402,15 @@ public struct SemanticAnalyzer: ASTPass {
   }
 
   public func postProcess(structDeclaration: StructDeclaration, passContext: ASTPassContext) -> ASTPassResult<StructDeclaration> {
-    return ASTPassResult(element: structDeclaration, diagnostics: [], passContext: passContext)
+    var diagnostics = [Diagnostic]()
+    // Detect Recursive types
+    let structName = structDeclaration.identifier.name
+
+    if let conflict = passContext.environment!.selfReferentialProperty(in: structName, enclosingType: structName) {
+      diagnostics.append(.recursiveStruct(structDeclaration.identifier, conflict))
+    }
+
+    return ASTPassResult(element: structDeclaration, diagnostics: diagnostics, passContext: passContext)
```

Comment:
Any reason this code is in `postProcess`? Otherwise, it should be in `process`

Reply:
My original thought was referencing types that hadn't been seen yet, but given the AST is already built this isn't actually an issue. So moved.